flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/CoordinatorEventsExactlyOnceTest.java

```diff
@@ -0,0 +1,350 @@
+import org.apache.flink.runtime.jobgraph.OperatorID;
+import org.apache.flink.runtime.operators.coordination.CoordinatorEventsExactlyOnceITCase;
+import org.apache.flink.runtime.operators.coordination.OperatorCoordinator;
+import org.apache.flink.runtime.operators.coordination.OperatorEvent;
+import org.apache.flink.runtime.operators.coordination.OperatorEventHandler;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.runtime.testutils.MiniClusterResource;
+import org.apache.flink.runtime.testutils.MiniClusterResourceConfiguration;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.sink.DiscardingSink;
+import org.apache.flink.streaming.api.functions.source.ParallelSourceFunction;
+import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
+import org.apache.flink.streaming.api.graph.StreamConfig;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperatorFactory;
+import org.apache.flink.streaming.api.operators.CoordinatedOperatorFactory;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperatorFactory;
+import org.apache.flink.streaming.api.operators.Output;
+import org.apache.flink.streaming.api.operators.StreamOperator;
+import org.apache.flink.streaming.api.operators.StreamOperatorParameters;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.util.Preconditions;
+import org.apache.flink.util.SerializedValue;
+
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.apache.flink.runtime.operators.coordination.CoordinatorEventsExactlyOnceITCase.EndEvent;
+import static org.apache.flink.runtime.operators.coordination.CoordinatorEventsExactlyOnceITCase.EventSendingCoordinator;
+import static org.apache.flink.runtime.operators.coordination.CoordinatorEventsExactlyOnceITCase.IntegerEvent;
+import static org.apache.flink.runtime.operators.coordination.CoordinatorEventsExactlyOnceITCase.StartEvent;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Integration test case that validates the exactly-once mechanism for operator events sent around
+ * checkpoint. This class is an extension to {@link
+ * org.apache.flink.runtime.operators.coordination.CoordinatorEventsExactlyOnceITCase}, further
+ * verifying the exactly-once semantics of events when the flink job is constructed using actual
+ * stream operators.
+ */
+public class CoordinatorEventsExactlyOnceTest {
```

Comment:
Could you explain why we add a new class here instead of adding tests in `CoordinatorEventsExactlyOnceITCase`?

If we do need to add new class, would you update class names and doc for clarify the difference?

In general we want the test class name to self-explain its purpose. It is not clear what is the difference between `CoordinatorEventsExactlyOnceITCase` and `CoordinatorEventsExactlyOnceTest`, and which class should be used for new tests in the future.

Reply:
Because the implementations of this PR are in `flink-runtime` and `flink-streaming-java`. Given that `flink-streaming-java` depends on `flink-runtime`, an e2e test for the problem solved in this PR should be in `flink-streaming-java`.

As already explained in the JavaDoc, The newly added test class verifies the correctness of coordinator events with actual stream operators, while the existing `CoordinatorEventsExactlyOnceITCase` makes the recipient of coordinator events a general `AbstractInvokable` subclass. The scope of these two test classes is different.

How do you like it to remove `CoordinatorEventsExactlyOnceITCase`? I now think that the `AbstractInvokable` may not provide more verifications than a `StreamOperator`, so the test case can be removed. If we agree on this idea, there will not be two integration tests in two packages about a similar situation, and the naming problem would not exist.

Reply:
According to our offline discussion, I'll give a more meaningful name to this test case and add corresponding descriptions in JavaDocs.